torch/lib/c10d/logger.cpp

```diff
@@ -85,7 +91,148 @@ void Logger::set_construction_data_and_log(
+  // end time can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  TORCH_CHECK(num_iterations_stats_recorded_ > 0);
+  if (cpu_end_time < cpu_start_time) {
+    return;
+  }
+  avg_time = ((cpu_end_time - cpu_start_time) +
+              avg_time * (num_iterations_stats_recorded_ - 1)) /
+      num_iterations_stats_recorded_;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    at::cuda::CUDAEvent& gpu_start,
+    at::cuda::CUDAEvent& gpu_end) {
+  TORCH_CHECK(num_iterations_stats_recorded_ > 0);
+  float milliseconds = gpu_start.elapsed_time(gpu_end);
+  // If gpu_end is not recorded in this iteration,
+  // milliseconds will have invalid value.
+  // For some cases like DDP runs on non-sync mode,
+  // gpu_end can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  if (milliseconds < 0) {
+    return;
+  }
+  avg_time = (int(milliseconds * kMilliSecondToNanosSecond) +
+              avg_time * (num_iterations_stats_recorded_ - 1)) /
+      num_iterations_stats_recorded_;
+}
+#endif
+
+void Logger::set_runtime_stats() {
+  // Sync with reducer's data
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
+  // set runtime stats after 1st iteration is complete.
+  if (!reducer_->should_collect_runtime_stats()) {
+    return;
+  }
+  num_iterations_stats_recorded_++;
+  // Set ith iteration when the runtime stats are set.
+  ddp_logging_data_->iteration = reducer_->num_iterations_;
+  // If unused_parameters_ is not empty, calculate its sizes.
+  // unused_parameters_ is calculated in forward call of
+  // each iteration.
+  for (const auto& unused_index : reducer_->unused_parameters_) {
+    const auto& v = reducer_->replicas_[unused_index.replica_index]
+                                       [unused_index.variable_index];
+    ddp_logging_data_->unused_parameter_size += v.numel() * v.element_size();
+  }
+  // rebuilt_bucket_sizes will not change once buckets are rebuilt,
+  // so it only needs to set once during whole training loop.
+  if (ddp_logging_data_->has_rebuilt_buckets != reducer_->has_rebuilt_bucket_) {
+    ddp_logging_data_->has_rebuilt_buckets = reducer_->has_rebuilt_bucket_;
+    ddp_logging_data_->rebuilt_bucket_sizes = get_bucket_sizes();
+  }
+
+  if (reducer_->replicas_[0][0].is_cuda()) {
+#ifdef USE_CUDA
+    // Cuda time stats are only collected for single process single
+    // device and single device module.
+    if (reducer_->replicas_.size() > 1 || reducer_->is_multi_device_module_) {
+      TORCH_WARN_ONCE(
+          "Cuda time stats are not collected for single process "
+          "multiple device program or multi-device modules.");
+      return;
+    }
+    // Check events on the replicas_[0][0].device().
+    at::cuda::set_device(reducer_->replicas_[0][0].device().index());
```

Comment:
will this change the current device beyond this scope? Or should we use a device guard instead?

Reply:
yes, that is also what I am concerned. I tried to use deviceGuard(), it hanged!  But yesterday I thought about it again, it hanged possibly because I did not set_device before synchronizing in the logger. 

so I tried to use deviceGuard again before recording and synchronizing, now it works fine!